File: C#.Net/JDI/Web/Selenium/Elements/Complex/table/Cell.cs
using System;
using System.Linq;
using Epam.JDI.Commons;
using Epam.JDI.Core;
using Epam.JDI.Web.Selenium.Base;
using Epam.JDI.Web.Selenium.DriverFactory;
using Epam.JDI.Web.Selenium.Elements.Base;
using Epam.JDI.Web.Selenium.Elements.Complex.table.interfaces;
using Epam.JDI.Web.Settings;
using OpenQA.Selenium;
using static System.String;
using static Epam.JDI.Core.Settings.JDISettings;

namespace Epam.JDI.Web.Selenium.Elements.Complex.table
{
    public class Cell : SelectableElement, ICell
    {
        public int RowIndex { set; get; }
        public int ColumnIndex { set; get; }
        public Table Table { get; set; }
        public int ColumnNum { get; set; }
        public int RowNum { get; set; }
        private string _columnName { get; set; }
        private string _rowName { get; set; }
        private readonly By _cellLocatorTemplate = By.XPath(".//tr[{1}]/td[{0}]");
        

        public Cell(int columnNum, int rowNum, string colName, string rowName,
                    By cellLocatorTemplate, Table table, int columnIndex = -1, int rowIndex = -1, IWebElement webElement = null)
        {
            if (columnIndex > 0)
                ColumnIndex = table.Rows.HasHeader && table.Rows.LineTemplate == null 
                    ? ColumnIndex + 1 
                    : ColumnIndex;
            WebElement = webElement;
            RowIndex = rowIndex;
            ColumnNum = columnNum;
            RowNum = rowNum;
            _columnName = colName;
            _rowName = rowName;
            if (cellLocatorTemplate != null)
                _cellLocatorTemplate = cellLocatorTemplate;
            Table = table;
            ClickAction = c => ((Cell)c).Get().Click();
    }
        
        public string ColumnName => _columnName != null && !_columnName.Equals("")
                    ? _columnName
                    : Table.Columns.Headers[ColumnNum - 1];

        public string RowName => _rowName != null && !_rowName.Equals("")
                    ? _rowName
                    : Table.Rows.Headers[RowNum - 1];
        protected Func<Cell, string> TextAction => c => Get().Text;


        protected new Func<Cell, bool> SelectedAction => c => Get().Selected;

        public SelectableElement Get()
        {
            return WebElement != null
                    ? new SelectableElement(webElement: WebElement)
                    : new SelectableElement(_cellLocatorTemplate.FillByMsgTemplate(ColumnIndex, RowIndex));
        }

        public T Get<T>(Type clazz) where T : WebBaseElement
        {
            T instance;
            try
            {
                instance = (T) Activator.CreateInstance(clazz.IsInterface
                        ? MapInterfaceToElement.ClassFromInterface(clazz)
                        : clazz);
            }
            catch
            {
                throw Exception("Can't get Cell from interface/class: " + clazz.ToString().Split("\\.").Last());
            }
            return Get(instance);
        }

        public T Get<T>(T cell) where T : WebBaseElement
        {
            var locator = cell.Locator;
            if (locator == null || locator.ToString().Equals(""))
                locator = _cellLocatorTemplate;
            if (!locator.ToString().Contains("{0}") || !locator.ToString().Contains("{1}"))
                throw Exception("Can't create cell with locator template " + cell.Locator
                        + ". Template for Cell should contains '{0}' - for column and '{1}' - for row indexes.");
            cell.WebAvatar.ByLocator = locator.FillByMsgTemplate(RowIndex, ColumnIndex);
            cell.Parent = Table;
            return cell;
        }

        public Cell UpdateData(string colName, string rowName)
        {
            if (IsNullOrEmpty(_columnName) && !IsNullOrEmpty(colName))
                _columnName = colName;
            if (IsNullOrEmpty(_rowName) && !IsNullOrEmpty(rowName))
                _rowName = rowName;
            return this;
        }
    }
}

File: Java/JDI/jdi-uitest-win/src/main/java/com/epam/jdi/uitests/win/winium/elements/complex/table/Cell.java
package com.epam.jdi.uitests.win.winium.elements.complex.table;
/*
 * Copyright 2004-2016 EPAM Systems
 *
 * This file is part of JDI project.
 *
 * JDI is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * JDI is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with JDI. If not, see <http://www.gnu.org/licenses/>.
 */


import com.epam.jdi.uitests.core.interfaces.MapInterfaceToElement;
import com.epam.jdi.uitests.core.interfaces.base.ISelect;
import com.epam.jdi.uitests.win.winium.elements.BaseElement;
import com.epam.jdi.uitests.win.winium.elements.WebCascadeInit;
import com.epam.jdi.uitests.win.winium.elements.base.SelectElement;
import com.epam.jdi.uitests.win.winium.elements.complex.table.interfaces.ICell;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;

import static com.epam.commons.LinqUtils.last;
import static com.epam.jdi.uitests.core.settings.JDISettings.exception;
import static com.epam.jdi.uitests.win.winium.driver.WebDriverByUtils.fillByMsgTemplate;

/**
 * Created by 12345 on 25.10.2014.
 */

/**
 * Cell have some info about its position:<br>
 * ColumnName, RowName, ColumnIndex, RowIndex<br>
 * You can do some actions with Cell:<br>
 * Click, Select, getText, waitText, waitMatchText<br>
 * Also you can use get() method to get Element of specified for table Type and do any possible action with it<br>
 */
class Cell extends SelectElement implements ISelect, ICell {
    private int rowIndex;
    private int columnIndex;
    private Table table;
    private int columnNum;
    private WebElement webElement;
    private int rowNum;
    private String columnName;
    private String rowName;
    private By cellLocatorTemplate = By.xpath(".//tr[{1}]/td[{0}]");

    Cell(WebElement webElement, int columnNum, int rowNum, String colName, String rowName,
         By cellLocatorTemplate, Table table) {
        this.webElement = webElement;
        this.columnNum = columnNum;
        this.rowNum = rowNum;
        this.columnName = colName;
        this.rowName = rowName;
        if (cellLocatorTemplate != null)
            this.cellLocatorTemplate = cellLocatorTemplate;
        this.table = table;
    }

    Cell(int columnIndex, int rowIndex, int columnNum, int rowNum, String colName, String rowName,
         By cellLocatorTemplate, Table table) {
        this.columnIndex = (table.rows().hasHeader && table.rows().lineTemplate == null) ? columnIndex + 1 : columnIndex;
        this.rowIndex = rowIndex;
        this.columnNum = columnNum;
        this.rowNum = rowNum;
        this.columnName = colName;
        this.rowName = rowName;
        if (cellLocatorTemplate != null)
            this.cellLocatorTemplate = cellLocatorTemplate;
        this.table = table;
    }

    public void setWebElement(WebElement webElement) {
        this.webElement = webElement;
    }

    public int columnNum() {
        return columnNum;
    }

    public int rowNum() {
        return rowNum;
    }

    public String columnName() {
        return (columnName != null && !columnName.equals(""))
                ? columnName
                : table.columns().headers().get(columnNum - 1);
    }

    public String rowName() {
        return (rowName != null && !rowName.equals(""))
                ? rowName
                : table.rows().headers().get(rowNum - 1);
    }

    @Override
    protected String getTextAction() {
        return get().getText();
    }

    @Override
    protected void clickAction() {
        get().click();
    }

    @Override
    protected boolean isSelectedAction() {
        return get().isSelected();
    }

    public SelectElement get() {
        SelectElement cell = webElement != null
                ? new SelectElement(webElement)
                : new SelectElement(fillByMsgTemplate(cellLocatorTemplate, columnIndex, rowIndex));
        cell.setParent(table);
        return cell;
    }

    public <T extends BaseElement> T get(Class<T> clazz) {
        T instance;
        try {
            instance = (clazz.isInterface())
                    ? (T) MapInterfaceToElement.getClassFromInterface(clazz).newInstance()
                    : clazz.newInstance();
        } catch (Exception ex) {
            throw exception("Can't get Cell from interface/class: " + last((clazz + "").split("\\.")));
        }
        return get(instance);
    }

    public <T extends BaseElement> T get(T cell) {
        By locator = cell.getLocator();
        if (locator == null || locator.toString().equals(""))
            locator = cellLocatorTemplate;
        if (!locator.toString().contains("{0}") || !locator.toString().contains("{1}"))
            throw exception("Can't create cell with locator template " + cell.getLocator()
                    + ". Template for Cell should contains '{0}' - for column and '{1}' - for row indexes.");
        cell.getAvatar().byLocator = fillByMsgTemplate(locator, columnIndex, rowIndex);
        cell.setParent(table);
        new WebCascadeInit().initElements(cell, avatar.getDriverName());
        return cell;
    }

    public Cell updateData(String colName, String rowName) {
        if ((columnName == null || columnName.equals("")) && !(colName == null || colName.equals("")))
            columnName = colName;
        if ((this.rowName == null || this.rowName.equals("")) && !(rowName == null || rowName.equals("")))
            this.rowName = rowName;
        return this;
    }
}